util/collections.py

```diff
@@ -0,0 +1,344 @@
+# -*- coding: utf-8 -*-
+import os
+import csv
+import json
+
+from pathlib import Path
+from functools import partial
+from util.helpers import MEGABYTE, GIGABYTE
+from util.audio import Sample, AUDIO_TYPE_WAV, AUDIO_TYPE_OPUS, LOADABLE_FILE_FORMATS
+
+FILE_EXTENSION_CSV = '.csv'
+FILE_EXTENSION_SDB = '.sdb'
+
+BIG_ENDIAN = 'big'
+INT_SIZE = 4
+BIGINT_SIZE = 2 * INT_SIZE
+MAGIC = b'SAMPLEDB'
+
+BUFFER_SIZE = 1 * MEGABYTE
+CACHE_SIZE = 1 * GIGABYTE
+
+SCHEMA_KEY = 'schema'
+CONTENT_KEY = 'content'
+MIME_TYPE_KEY = 'mime-type'
+MIME_TYPE_TEXT = 'text/plain'
+CONTENT_TYPE_SPEECH = 'speech'
+CONTENT_TYPE_TRANSCRIPT = 'transcript'
+META = {
+    SCHEMA_KEY: [
+        {CONTENT_KEY: CONTENT_TYPE_SPEECH, MIME_TYPE_KEY: AUDIO_TYPE_OPUS},
+        {CONTENT_KEY: CONTENT_TYPE_TRANSCRIPT, MIME_TYPE_KEY: MIME_TYPE_TEXT}
+    ]
```

Comment:
Why is the schema a list instead of a dictionary? Do we support multiple "columns" of the same type? From reading the code it seems like in that case it would just use the first column of a given type and ignore the others. Turning this into a dict would simplify the code that consumes this data.

Reply:
- Up until recently dictionaries were not guaranteeing a certain order. Yes, this could be fixed by alphabetically ordering the keys, but this would introduce a similar indirection as the current one. 
- Breaking conceptional symmetry (schema using dict, while rows being lists).
- Current approach allows multiple representations / mime-types for the same semantic key.

Reply:
But I am open for changing it...

Reply:
Just tried a clean implementation of the dictionary approach and came to the conclusion that the current solution still feels better.

Reply:
Could you rephrase those three points in more concrete terms? I don't understand the implications for our use case specifically. For example, why is ordering needed here?

Also, for the third point, you didn't answer my original question: do we support multiple representations of the same type at all?

SDB file has meta info in it:

```
self.meta = {
  'schema': [
    {'content': 'speech', 'mime-type': 'audio/opus'},
    {'content': 'transcript', 'mime-type': 'text/plain'},
  ],
}
```

Then it's read like this:

```python
self.schema = self.meta['schema']
self.speech_index = self.find_column(content='speech')
self.audio_type = self.schema[self.speech_index]['mime-type']
self.transcript_index = self.find_column(content='transcript')
self.text_type = self.schema[self.transcript_index]['mime-type']
```

And `find_column` is:

```python
def find_column(self, content):
  for index, column in enumerate(self.schema)
      if column['content'] == content:
        return index
  return -1
```

This means only the first entry for a given `'content'` will ever be read. So we're effectively reading this as a dictionary where a key can only have one value, but it's represented as a list of tuples. Maybe add a warning or error message if more than one entry is found for a given key? That would at least avoid people thinking we support more than we do.

Reply:
The schema represents the order of items in one DB "row".
The third point is for future compatibility but not as important.

Reply:
Ah, I was completely missing the fact that the order in the schema follows the order in the binary encoding. So maybe just add a comment or a warning message if schema 'content' keys appear more than once?

Reply:
It is implemented as `dict` now.